### PR TITLE
(GH-238) Add STDIO support

### PR DESCRIPTION
**Pull request by jpogran**
This PR adds STDIO support to the extension

This build on PR #272 so it must be merged first.

### CODE REVIEW

File: src/configuration.ts [hunk 1]
```diff
@@ -94,60 +98,86 @@ export class ConnectionConfiguration implements IConnectionConfiguration {
...
-      args.push('--ip=127.0.0.1');
-    } else {
-      args.push('--ip=' + this.host);
+    switch (this.protocol) {
+      case ProtocolType.STDIO:
+        args.push('--stdio');
+        break;
+      case ProtocolType.TCP:
+        if (this.host === undefined || this.host === '') {
+          args.push('--ip=127.0.0.1');
```
Comment (glennsarti) [inline src/configuration.ts]: might be best to change this to `localhost`.  Supports ip6 and ip4 then

Comment (jpogran) [reply src/configuration.ts]: in a PR about STDIO?

Comment (glennsarti) [reply src/configuration.ts]: whoops....True...just saw the green

**Comment by jpogran** [comment]
@glennsarti @michaeltlombardi can you test this this week and validate it works for you?

**Comment by glennsarti** [comment]
The STDIO transport works
The Debug Server still works

But I'm having trouble with the TCP transport.

If I only set the protocol to tcp
```
"puppet.languageclient.protocol": "tcp",
```
Then the extension get's stuck on starting puppet

The extension logs show;
```
DEBUG: 2018-06-13T12:16:39.995Z Language server PID:14684
DEBUG: 2018-06-13T12:16:39.996Z LanguageServer Process Started: 14684
DEBUG: 2018-06-13T12:16:40.843Z pdk shell started, pid: 9388
DEBUG: 2018-06-13T12:16:44.281Z OUTPUT: LANGUAGE SERVER RUNNING 127.0.0.1:51439

DEBUG: 2018-06-13T12:16:44.282Z Starting TCP client: 
DEBUG: 2018-06-13T12:16:54.182Z Configuring language server options
DEBUG: 2018-06-13T12:16:54.183Z Starting language server client (host 127.0.0.1 port 51439)
DEBUG: 2018-06-13T12:16:54.184Z Configuring language server client options
```

I added a debug log to the lang. server and it's definitely starting and then timing out after 10 seconds because there's no connection.

I also added one line for debugging purposes
``` typscript
          if (this.languageServerClient === undefined && data.toString().match('LANGUAGE SERVER RUNNING') !== null) {
            this.logger.debug('Starting TCP client: ');  //  <----------- I ADDED THIS LINE
            this.languageServerClient = this.startLanguageClient();
          }

```

I then noticed the timestamps. 10 second delay.  I think we have a blocking process in there somewhere
```
DEBUG: 2018-06-13T12:16:44.282Z Starting TCP client: 
DEBUG: 2018-06-13T12:16:54.182Z Configuring language server options
```

**Comment by jpogran** [comment]
@glennsarti Fixed

File: src/connection.ts [hunk 2]
```diff
@@ -116,15 +121,24 @@ export class ConnectionManager implements IConnectionManager {
...
+          if (this.languageServerClient === undefined && data.toString().match('LANGUAGE SERVER RUNNING') !== null) {
+            this.languageServerClient = this.createLanguageClient();
+            this.extensionContext.subscriptions.push(this.languageServerClient.start());
+          }
+        });
+        break;
+      case ProtocolType.STDIO:
+        this.logger.debug('Starting STDIO client: ');
+        this.languageServerClient = this.createLanguageClient();
         this.extensionContext.subscriptions.push(this.languageServerClient.start());
```
Comment (jpogran) [inline src/connection.ts]: I tried to have this line not duplicated by moving it to line 142. However due to closures and how typescript/node/javacript/whatever works, line 142 is hit before anything in the case statement is evaluated and thus a null client is added before its actually initialized.

**Pull request merged**